common/Resources/ti.internal/extensions/util.js

```diff
@@ -0,0 +1,497 @@
+	isPrimitive: value => (typeof value !== 'object' && typeof value !== 'function') || value === null,
+	isString: value => typeof value === 'string',
+	isSymbol: value => typeof value === 'symbol',
+	isUndefined: value => value === undefined,
+	log: string => {
+		const date = new Date();
+		const time = `${date.getHours().toString().padStart(2, '0')}:${date.getMinutes().toString().padStart(2, '0')}:${date.getSeconds().toString().padStart(2, '0')}`;
+		// Produces output like: "21 Feb 10:04:23 - message"
+		console.log(`${date.getDate()} ${MONTHS[date.getMonth()]} ${time} - ${string}`);
+	},
+	print: (...args) => console.log(args.join('')), // FIXME: Shouldn't add trailing newline like console.log does!
+	puts: (...args) => console.log(args.join('\n')),
+	error: (...args) => console.error(args.join('\n')),
+	debug: string => console.error(`DEBUG: ${string}`)
+};
+
+util.isBuffer = () => false; // FIXME: Check for Ti.Buffer? for node/browserify buffer?
+util.isDate = value => util.types.isDate(value);
+util.isError = value => util.types.isNativeError(value);
+util.isRegexp = value => util.types.isRegexp(value);
+
+// FIXME: Our String.format is not very forgiving. It sort-of is supposed to do the same thing, but blows up easily
+// util.format = String.format;
+function getConstructor(obj) {
+	if (obj.constructor) {
+		return obj.constructor.name;
+	}
+	return 'Object';
+}
+
+const defaultInspectOptions = {
+	showHidden: false,
+	depth: 2,
+	colors: false,
+	customInspect: true,
+	showProxy: false,
+	maxArrayLength: 100,
+	breakLength: 60,
+	compact: true,
+	sorted: false,
+	getters: false
+};
+
+/**
+ * @param {*} obj JS value or object to inspect
+ * @param {object} [options] options for output
+ * @param {Integer} [options.breakLength=60] length at which to break properties into individual lines
+ * @param {boolean} [options.showHidden=false] whether to include hidden properties (non-enumerable)
+ * @param {boolean} [options.sorted=false] whether to sort the property listings per-object
+ * @param {boolean} [options.compact=true] if set to `false`, uses luxurious amount of spacing and newlines
+ * @param {Integer} [options.depth=2] depth to recurse into objects
+ * @returns {string}
+ */
+util.inspect = (obj, options = {}) => {
+	const mergedOptions = Object.assign({}, defaultInspectOptions, options);
+	// increase our recursion counter to avoid going past depth
+	if (mergedOptions.recursionCount === undefined) {
+		mergedOptions.recursionCount = -1;
+	}
+	mergedOptions.recursionCount++;
+	if (mergedOptions.indentLevel === undefined) {
+		mergedOptions.indentLevel = 0;
+	}
+	try {
+		const objType = typeof obj;
+		if (objType === 'object' || objType === 'function') {
+			if (obj === null) {
+				return 'null';
+			}
+
+			// Guard against circular references
+			// FIXME: Need to push/pop the references, not store forever!
+			mergedOptions.memo = mergedOptions.memo || [];
+			if (mergedOptions.memo.includes(obj)) {
```

Comment:
What about repeated occurrences of the same object, for example: `const a={id:1}; util.inspect([a,a]);`

Reply:
good catch, and yes that test is a nice way to check. I've since added this as a unit test and fixed it.